src/plugins/data/common/query/persistable_state.ts

```diff
@@ -6,58 +6,53 @@
  * Side Public License, v 1.
  */
 
-import uuid from 'uuid';
-import { Filter } from '@kbn/es-query';
 import { SavedObjectReference } from '@kbn/core/types';
-import { MigrateFunctionsObject } from '@kbn/kibana-utils-plugin/common';
-import { DATA_VIEW_SAVED_OBJECT_TYPE } from '..';
+import { MigrateFunctionsObject, VersionedState } from '@kbn/kibana-utils-plugin/common';
+import type { QueryState } from './query_state';
+import * as filtersPersistableState from './filters/persistable_state';
 
-export const extract = (filters: Filter[]) => {
+export const extract = (queryState: QueryState) => {
   const references: SavedObjectReference[] = [];
-  const updatedFilters = filters.map((filter) => {
-    if (filter.meta?.index) {
-      const id = uuid();
-      references.push({
-        type: DATA_VIEW_SAVED_OBJECT_TYPE,
-        name: id,
-        id: filter.meta.index,
-      });
 
-      return {
-        ...filter,
-        meta: {
-          ...filter.meta,
-          index: id,
-        },
-      };
-    }
-    return filter;
-  });
-  return { state: updatedFilters, references };
+  const { state: updatedFilters, references: referencesFromFilters } =
+    filtersPersistableState.extract(queryState.filters ?? []);
+  references.push(...referencesFromFilters);
+
+  return {
+    state: {
+      ...queryState,
+      filters: updatedFilters,
+    },
+    references,
+  };
 };
 
-export const inject = (filters: Filter[], references: SavedObjectReference[]) => {
-  return filters.map((filter) => {
-    if (!filter.meta.index) {
-      return filter;
-    }
-    const reference = references.find((ref) => ref.name === filter.meta.index);
-    return {
-      ...filter,
-      meta: {
-        ...filter.meta,
-        index: reference && reference.id,
-      },
-    };
-  });
+export const inject = (queryState: QueryState, references: SavedObjectReference[]) => {
+  const updatedFilters = filtersPersistableState.inject(queryState.filters ?? [], references);
+
+  return {
+    ...queryState,
+    filters: updatedFilters,
+  };
 };
 
-export const telemetry = (filters: Filter[], collector: unknown) => {
-  return {};
+export const telemetry = (queryState: QueryState, collector: unknown) => {
+  const filtersTelemetry = filtersPersistableState.telemetry(queryState.filters ?? [], collector);
+  return {
+    ...filtersTelemetry,
+  };
 };
 
-export const migrateToLatest = (filters: Filter[], version: string) => {
-  return filters;
+export const migrateToLatest = ({ state, version }: VersionedState<QueryState>) => {
+  const migratedFilters = filtersPersistableState.migrateToLatest({
+    state: state.filters ?? [],
+    version,
+  });
+
+  return {
+    ...state,
+    filters: migratedFilters,
+  };
 };
 
 export const getAllMigrations = (): MigrateFunctionsObject => {
```

Comment:
we should getAllMigrations from filters here

Reply:
Please take a look https://github.com/elastic/kibana/pull/132623/commits/ef7f2da3621c91c9fdc74dc3f46f04f098e15832
I followed the search source example, but not sure about `filter` vs `filters`. seems like search source example is wrong